Dev objectoriented - Moodle 3.5+

bradpasley opened the pull request.

Commit:
Refactor into classes

Commit:
add @template tag to template files

Commit:
fix header comment

Commit:
wrap long lines

Commit:
correct indenting

Commit:
improve diagnosis phpunit test, commenting, css

Commit:
update mermaid: remove unneeded attrib. for surgeon

Commit:
Adjust github workflow to mdl35 (main) branch

Commit:
add class documentation and remove attr class type

Commit:
fix example context json

Commit:
improve documenting of classes, constructors and constants

Commit:
include Moodle 3.7 in this branch (M3.5+)

Commit:
correct documentation of deletetasklist

Commit:
Adjust support range of branch

Commit:
merge to one branch to cover Moodle 3.5+

Commit:
Merge branch 'main' into dev-objectoriented-35

golenkovm commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,11 @@
+    uses: catalyst/catalyst-moodle-workflows/.github/workflows/ci.yml@main
+    # secrets:
+    # Required if you plan to publish (uncomment the below)
+    #  moodle_org_token: ${{ secrets.MOODLE_ORG_TOKEN }}
```
An empty line is missing at the end if the file

golenkovm commented on README.md:
```diff
@@ -39,11 +46,9 @@ Instructions of how to use the CLI can be found in the help page:
-Moodle 4.0+        | MOODLE_400_STABLE
+| Moodle version     | Branch            | PHP       |
+-------------------- | ------------------|-----------|
+| Moodle 3.5+        | main              | 7.1 - 7.4 |
```
It seems like `PHP` column doesn't add any value here. It should either specify which Moodle version requires which PHP version, but as there is only one branch I would just remove the column.

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+ * @package     tool_fix_delete_modules
+ * @category    admin
+ * @author      Brad Pasley <brad.pasley@catalyst-au.net>
+ * @copyright   Catalyst IT, 2022
```
`@copyright` tags here and in all other files are expected to be like `@copyright   2022 Catalyst IT`. See https://docs.moodle.org/dev/Coding_style#.40copyright

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    /**
+     * get_contextid() - Get the name of the table related to the course module which is failing to be deleted.
```
1. I believe here in other places method name can be removed from the docs.
2. _Get the name of the table_ doesn't match `return $this->modulecontextid;`

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * get_contextid() - Get the name of the table related to the course module which is failing to be deleted.
+     *
+     * @return string
```
Shouldn't the type be `int`?

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+
+
+    /**
+     * get_modulename() - Get the name of the table related to the course module which is failing to be deleted.
```
Seems like a copy&paste issue: `Get the name of the table...`

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+        if (isset($coursemoduleid)) {
+            global $DB;
+
+            if ($result = $DB->get_records('context', array('contextlevel' => '70', 'instanceid' => $coursemoduleid))) {
```
Have you considered using core `context_module::instance($cm->id);` here?

bradpasley replied:
Using that will throw a dml_missing_record_exception in some scenarios. But I can use it in a try/catch (and the catch would excecute the current code)

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    /**
+     * set_modulename()-  Set the name of the table related to the course module which is failing to be deleted.
```
Please feel free to review all the docs

golenkovm replied:
I mean it sets module name, not the table

golenkovm replied:
Seems to be unresolved yet

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+     * @param int $coursemoduleid - coursemodule id.
+     * @param int $moduleinstanceid - moduleinstance id.
+     *
+     * @return null
```
It should be `@return void`

golenkovm replied:
Seems to be unresolved yet

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+            $queryarray = array('id' => $coursemoduleid, 'instance' => $moduleinstanceid);
+        }
+
+        if ($cmrecord = $DB->get_records('course_modules', $queryarray, '', 'module')) {
```
Can this be replaced by core function `get_coursemodule_from_id($modulename, $cmid);` with empty `$modulename`?

bradpasley replied:
This would fail to obtain the modulename when the module's table (e.g. quiz) record was unexpectedly absent. Whereas direct DB query without JOINing to module's table will still get the modulename.
This is due to this query in get_coursemodule_from_id():
https://github.com/moodle/moodle/blob/master/lib/datalib.php#L1257

golenkovm replied:
Agreed

golenkovm commented on classes/diagnosis.php:
```diff
@@ -0,0 +1,151 @@
+    public function __construct(delete_task $task, array $symptoms) {
+        $stringtmm = get_string($this::TASK_MULTIMODULE, 'tool_fix_delete_modules');
+        $stringtam = get_string($this::TASK_ADHOCRECORDMISSING, 'tool_fix_delete_modules');
+        $stringmrm = get_string($this::MODULE_MODULERECORDMISSING, 'tool_fix_delete_modules');
```
What is the benefit of using it as a constant? As I can see it's used for strings only, so can be safely replaced with `'tool_fix_delete_modules'` across the code.
Same for other string constants.

bradpasley replied:
Could you clarify? are you saying to use `get_string($this::TASK_MULTIMODULE, 'tool_fix_delete_modules())` directly instead of `$striungmm` or not using the constants e.g. `$this::TASK_MULTIMODULE`?
I've used constants to help with consistency of the strings being used (even between classes).

golenkovm replied:
I mean not using the constants.
Just use `get_string('symptom_multiple_modules_in_task', 'tool_fix_delete_modules');`.
Personally, constants do not add any value here. Either not making the code more consistent. But instead we get more bulky code that is harder to follow.

golenkovm commented on classes/surgeon.php:
```diff
@@ -0,0 +1,376 @@
+        }
+        $outcomemessages[] = get_string(outcome::MODULE_SUCCESS, 'tool_fix_delete_modules');
+
+        return $outcomemessages;
```
This method seems to be returning inconsistent data types: in some cases it returns `outcome` object and in other cases it returns an array

bradpasley replied:
I'm not sure why it would be returning an outcome object. It should always be an array of outcome objects. Can you clarify more?

golenkovm replied:
You have `return new outcome($task, $outputmessages);` on the lines 193 and 201.
But here you return `$outcomemessages` which is array.

golenkovm commented on classes/surgeon.php:
```diff
@@ -0,0 +1,376 @@
+        if (!isset($returnname)) {
+            global $DB;
+            // If cm doesn't have the module id, try to retrieve that from the db.
+            $moduleid = isset($cm->module) ? $cm->module : null;
```
Why its checking `$cm->module` and not `$cm->id`?

bradpasley replied:
We're trying to get the modulename from the 'name' field in {modules}, so first need to get the 'module' value from {course_modules}

golenkovm replied:
Ok, all good

golenkovm commented on classes/reporter.php:
```diff
@@ -0,0 +1,747 @@
+     * @param array $querytaskids list of specific tasks to report/diagnose/fix (optional).
+     */
+    public function __construct(bool $ishtmloutput = true, int $minimumfaildelay = 60, array $querytaskids = array()) {
+        $this->ishtmloutput = $ishtmloutput;
```
This should be a separate object like `logger` or at least a private method like `format_message()` that would cook a message for you. It would simplify the code and get rid of multiple 
```
            if ($this->ishtmloutput) {
                ...
            } else {
                ...
            }
```

golenkovm commented on tests/test_fix_course_delete_module_class_deletemodule_test.php:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @coversNothing
+     */
+    public function test_deleting() {
```
Can't get what exactly are we testing here?

bradpasley replied:
Just ensuring that the database is working as expected.

golenkovm replied:
This seems to be out of scope for this plugin. I would suggest to remove this kind of unit tests here and in other test classes.

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return null
+     */
+    public function set_modulename(int $coursemoduleid, ?int $moduleinstanceid = null) {
```
Looks like this method should be private

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+     *
+     * @return null
+     */
+    public function set_contextid(int $coursemoduleid) {
```
Looks like this method should be private

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+        $this->courseid = $courseid;
+        $this->section = $section;
+        $this->set_contextid($this->coursemoduleid);
+        $this->set_modulename($this->coursemoduleid, $this->moduleinstanceid);
```
Why do you need to pass `coursemoduleid` and `moduleinstanceid` to `set_contextid()` and `set_modulename()` if they can be accessed as properties from the methods?

golenkovm commented on classes/deletemodule.php:
```diff
@@ -0,0 +1,144 @@
+                                ?int $moduleinstanceid = null,
+                                ?int $courseid = null,
+                                ?int $section = null) {
+        $this->taskid = $taskid;
```
Can't find how this property is used in the class

bradpasley replied:
agreed. will remove it.

golenkovm commented on tests/test_fix_course_delete_module_class_deletetask_test.php:
```diff
@@ -0,0 +1,209 @@
+     *
+     * @coversNothing
+     */
+    public function test_deleting() {
```
Can't get the idea of this unit test. The plugin does nothing here

bradpasley replied:
Just ensuring that the database is working as expected.

bradpasley replied:
Going to remove

golenkovm commented on tests/test_fix_course_delete_module_class_deletetask_test.php:
```diff
@@ -0,0 +1,209 @@
+        $this->assertCount(2, $coursedmodules);
+
+        // Remove previous adhoc task.
+        $this->assertCount(1, $DB->get_records('task_adhoc'));
```
You shouldn't rely on vanilla moodle behaviour only. Other plugins may change it via observers and create extra ad-hoc tasks (or any other data). See https://tracker.moodle.org/browse/MDL-62483

golenkovm commented on tests/test_fix_course_delete_module_class_deletetask_test.php:
```diff
@@ -0,0 +1,209 @@
+        try { // This will fail due to the quiz record already being deleted.
+            $removaltaskmulti->execute();
+        } catch (Exception $e) {
+            $this->assertCount(1, $DB->get_records('task_adhoc'));
```
`$this->expectException()` is more appropriate here as it will throw an exception if `$removaltaskmulti->execute();` doesn't fail for some reason.

golenkovm commented on classes/deletetask.php:
```diff
@@ -0,0 +1,196 @@
+class delete_task {
+    /** @var int $taskid - the course_delete_module adhoc task id for this course module. */
+    public $taskid;
+    /** @var array $deletemodules - an array of delete_module objects.*/
```
Here and in other places the best way to document a type as an array of a given class is `delete_module[]`. See https://docs.moodle.org/dev/Coding_style#Types

golenkovm commented on tests/test_fix_course_delete_module_class_deletetask_test.php:
```diff
@@ -0,0 +1,209 @@
+        $this->assertEquals([$assigncontextid, $quizcontextid], $deletetask->get_contextids());
+        $this->assertEquals([$assign->cmid => 'assign', $quiz->cmid => 'quiz'], $deletetask->get_modulenames());
+
+        // Execute tasks (first one should complete, second should fail).
```
I can see only one task being executed while comment says there are two tasks: _first one should complete, second should fail_

bradpasley replied:
Yes, that's not a helpful comment. will correct it to "Execute task (assign cm should complete, quiz cm should fail)."

golenkovm commented on tests/test_fix_course_delete_module_class_deletetasklist_test.php:
```diff
@@ -0,0 +1,201 @@
+        // Include all tasks (even 0 fail delay).
+        $deletetasklist = new delete_task_list(0);
+
+        $deletetasks        = array();
```
Redundant line?

bradpasley replied:
will remove

golenkovm replied:
Unresolving the conversation. Please feel free to resolve once this is fixed.

golenkovm commented on classes/deletetasklist.php:
```diff
@@ -0,0 +1,80 @@
+     * set_deletetasks() - Set the deletetasks array.
+     *
+     */
+    public function set_deletetasks() {
```
Please change the visibility to private

bradpasley replied:
probably a copy-paste caused mistake. will update

golenkovm commented on tests/test_fix_course_delete_module_class_deletetasklist_test.php:
```diff
@@ -0,0 +1,201 @@
+        $deletetasks        = array();
+        $deletetasks        = $deletetasklist->get_deletetasks();
+        $deletepagetask     = current($deletetasks);
+        $deletemultitask    = end($deletetasks);
```
Relying on the default db sorting might be dangerous and lead to unexpected order. You assume that multitask goes last but in real life db engine might return it first. Also there might be extra adhoc tasks generated by other plugins (eg on course or user creation)

golenkovm commented on tests/test_fix_course_delete_module_class_deletetasklist_test.php:
```diff
@@ -0,0 +1,201 @@
+        $this->assertEquals(end($dbtasks)->id, $dmquiz->taskid);
+        $this->assertEquals(end($dbtasks)->id, end($deletetasks)->taskid);
+
+        // Execute tasks (first one should complete, second should fail).
```
This comment is a bit confusing - only one task is being executed below

golenkovm commented on tests/test_fix_course_delete_module_class_diagnoser_test.php:
```diff
@@ -0,0 +1,236 @@
+        $deletetasks        = array_values($deletetasklist->get_deletetasks());
+        $deletepagetask     = $deletetasks[0];
+        $deleteurltask      = $deletetasks[1];
+        $deletemultitask    = $deletetasks[2];
```
Same potential issue here - sorting is not guaranteed

golenkovm commented on tests/test_fix_course_delete_module_class_diagnoser_test.php:
```diff
@@ -0,0 +1,236 @@
+        $removalbooktask->set_custom_data($bookdata);
+        $bookcms = $removalbooktask->get_custom_data();
+
+        // Execute tasks (first one should complete, second should fail).
```
This comment is chasing me hehe

golenkovm commented on tests/test_fix_course_delete_module_class_diagnoser_test.php:
```diff
@@ -0,0 +1,236 @@
+        $diagnosermultitask = new diagnoser($deletemultitask);
+        $diagnoserbooktask  = new diagnoser($deletebooktask);
+
+        $expectedsymptomspage  = [''.$page->cmid =>
```
What is `''.`?

bradpasley replied:
fixing...

golenkovm commented on tests/test_fix_course_delete_module_class_diagnoser_test.php:
```diff
@@ -0,0 +1,236 @@
+        // Check diagnoser for book deletion task (non-existant task).
+        $this->assertFalse($deletebooktask->is_multi_module_task());
+        $this->assertFalse($deletebooktask->task_record_exists());
+        $this->assertEquals($expecteddiagnosisbooktask, $diagnoserbooktask->get_diagnosis());
```
This test is really hard to follow. I would recommend to split it into a number of shorter and clearer tests. You can use `setUp()` method to create required data (course, cms, tasks etc)

bradpasley replied:
fixing...

golenkovm commented on tests/test_fix_course_delete_module_class_reporter_test.php:
```diff
@@ -0,0 +1,244 @@
+        $assign   = $this->getDataGenerator()->create_module('assign', array('course' => $course->id));
+        $assigncm = get_coursemodule_from_id('assign', $assign->cmid);
+        $quiz     = $this->getDataGenerator()->create_module('quiz', array('course' => $course->id));
+        $quizcm   = get_coursemodule_from_id('quiz', $quiz->cmid);
```
I can see that this block is present in (almost) every test class. You can create a base class for your tests and put this course and module stuff there to avoid excessive repetitions.

bradpasley replied:
fixing...

golenkovm commented on tests/test_fix_course_delete_module_class_outcome_test.php:
```diff
@@ -0,0 +1,169 @@
+
+        // Setup 2nd adhoc task for a multi-module delete (both quiz and assign).
+        $removaltaskmulti = new \core_course\task\course_delete_modules();
+        $cmsarray = array(''.$assigncm->id => array('id' => $assigncm->id),
```
Do you use `''.` to convert int to string?  Would `(string) $assigncm->id` work?

bradpasley replied:
fixing

golenkovm commented on tests/test_fix_course_delete_module_class_reporter_test.php:
```diff
@@ -0,0 +1,244 @@
+ * @copyright   Catalyst IT, 2022
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class test_fix_course_delete_module_class_reporter_test extends \advanced_testcase {
```
It seems like `get_tables_report()` is a critical public method but it's not covered by tests

golenkovm commented on delete_module.php:
```diff
@@ -50,24 +50,15 @@
-        }
-    }
-    $deleteoutput = force_delete_module_data($cm, $taskid, true);
+    $minimumfaildelay = intval(get_config('tool_fix_delete_modules', 'minimumfaildelay'));
```
Can this be passed down to reporter constructor?

bradpasley replied:
will fix this

golenkovm replied:
For the review convenience please feel free to resolve the issue once it's actually fixed. Otherwise, it's a bit hard to track review notes

golenkovm commented on delete_module.php:
```diff
@@ -24,12 +24,11 @@
+require_once(__DIR__.'/classes/reporter.php');
 require_login();
 
+use tool_fix_delete_modules\reporter as reporter;
```
Here and in other places can this be just `use tool_fix_delete_modules\reporter;`?

golenkovm commented on classes/reporter.php:
```diff
@@ -0,0 +1,747 @@
+        $deletetaskslist = new delete_task_list($this->minimumfaildelay);
+        $deletetasks = $deletetaskslist->get_deletetasks();
+        foreach ($deletetasks as $taskid => $deletetask) {
+            if (empty($taskids) || in_array($taskid, $taskids)) { // If no filter or is in the filter.
```
Is this method (and `make_fix()` ) being called anywhere with no filters/tasks set? If not I would just stick with `$taskids` as a required param and do nothing if it's not passed (or throw an exception).

bradpasley replied:
Yes, it is, in `get_diagnosis()`

golenkovm replied:
Could you please provide the link?

bradpasley replied:
https://github.com/catalyst/moodle-tool_fix_delete_modules/blob/dev-objectoriented-35/classes/reporter.php#L111

golenkovm commented on classes/reporter.php:
```diff
@@ -0,0 +1,747 @@
+     *
+     * @return string
+     */
+    public function make_fix(array $taskids = array()) {
```
It would make it clear to rename the method to something like `fix_tasks()`

Commit:
fix rescheduling of tasks

Commit:
working through improvements

bradpasley committed:
Merge Updates into Dev objectoriented 35 branch (#11)

* added improvements according to feedback
* improved diagnosis and reporter unittests
* add format_message() function, improve cli
* miniqueue only if it exists (in Moodle 4.1 only)

golenkovm commented on styles.css:
```diff
@@ -1,25 +1,98 @@
+    position: absolute;
+    left: -999px;
     max-width: 0;
 }
```
An empty line at the end of the file is missing here

golenkovm commented on tests/test_fix_course_delete_module_class_deletemodule_test.php:
```diff
@@ -0,0 +1,115 @@
+        $this->assertEmpty($DB->get_records('page'));
+        $this->assertEmpty($DB->get_records('book'));
+        $this->assertEmpty($DB->get_records('url'));
+        $this->assertEmpty($DB->get_records('task_adhoc'));
```
This test should be deleted as well. You can not guarantee these numbers. 3rd party plugins can throw their stuff on install eg create enrolments or post-install adhoc tasks. Also this test doesn't really test the `deletemodule` class.

golenkovm commented on tests/test_fix_course_delete_module_class_reporter_test.php:
```diff
@@ -0,0 +1,339 @@
+        // Test output displays for get_tables_report().
+        $testreports = $testreporter->get_tables_report();
+        $this->assertNotEquals('', $testreports);
+        $this->assertTrue(mb_strpos($testreports, get_string('report_heading', 'tool_fix_delete_modules')) !== false);
```
Wouldn't `strpos()` work here?

golenkovm commented:
Hi @bradpasley 
At the end of the review there will be another thing to do. We'll need to update class and test files so they meet https://docs.moodle.org/dev/Coding_style and https://docs.moodle.org/dev/Writing_PHPUnit_tests:
1. Class file names should match the class name, eg `classes/deletemodule.php` should either contain `deletemodule` class or be renamed to `classes/delete_module.php` (I prefer the second option).
2. Test files should have names like `classnametobetested_test.php`, eg `tests/delete_module_test.php` with `delete_module_test` class in it.

Please kindly wait for all the discussions to be resolved before proceeding with this one. Otherwise, it would be hard to track them.

bradpasley committed:
PHPUnit testing base class, update string castings (#13)

* miniqueue only if it exists (for Moodle master branch)
* int to string casts, base testing class, setUp()

Commit:
WIP add draft requeue and execute adhoctask functionality

Commit:
reschedule and execute adhoc tasks for singular module tasks

Commit:
execute individualised course_delete_module tasks right after creation

bradpasley committed:
Merge pull request #14 from catalyst/dev-requeue-asap-and-execute-adhoctask

Re-queue course_delete_module Adhoc Tasks asap and then execute them

Commit:
functions docs, reschedule unit tests, constants

- Reformat function documentation
- Add phpunit tests for reschedule_or_queue function
- Refactor constants into language strings

Commit:
fix return types of delete_module_cleanly()

Commit:
adjust css, return void, mb_strpos/strpos function

bradpasley committed:
Merge pull request #15 from catalyst/dev-requeu-asap-and-execute-adhoctask

functions docs, reschedule unit tests, constants

golenkovm commented on classes/surgeon.php:
```diff
@@ -0,0 +1,520 @@
+                $outcomemessages[] = get_string('outcome_adhoc_task_record_reexecution_failed', 'tool_fix_delete_modules');
+            }
+        } else {
+            $dbtasks = $DB->get_records('task_adhoc');
```
Do we need this line?

bradpasley replied:
removed

golenkovm commented on classes/surgeon.php:
```diff
@@ -0,0 +1,520 @@
+                $this->reschedule_or_queue_adhoc_task($thisadhoctask);
+            }
+            $outcomemessages[] = get_string('outcome_adhoc_task_record_rescheduled', 'tool_fix_delete_modules');
+            try {
```
I'm pretty sure we don't need to run the task in user session. We've rescheduled the task, re-set its faildelay and nextruntime and this is enough. Leave the rest to cron. As long as faildelay is re-set, this task should not be picked up by the plugin report, so we don't need to worry about the case when someone pushed the button twice for one task

golenkovm commented on tests/test_fix_course_delete_module_class_surgeon_test.php:
```diff
@@ -0,0 +1,328 @@
+        // Schedule the task again.
+        $task = $this->removaltaskassign;
+        $task->set_next_run_time($newruntime);
+        surgeon::reschedule_or_queue_adhoc_task($task);
```
It seems like these tests are valid for Moodle 3.5-3.6 only. `surgeon::reschedule_or_queue_adhoc_task()` is not supposed to be called in Moodle 3.7 onwards.
The best unit test here would be to fail the task, then fix it using public API and then assert that the task was rescheduled. Such unit test would work in differently in Moodle 3.5-3.6 (using `surgeon::reschedule_or_queue_adhoc_task()`) vs Moodle 3.7 onwards (`using cron API`), but the outcome will be the same and that's exactly what we need to test here.

bradpasley replied:
the function is refactored to call the `\core\task\manager::reschedule_or_queue_adhoc_task()` if it exists, otherwise very similar code will be called from within the surgeon class.
phpunittests confirm that tasks are rescheduled for all Moodle versions from 3.5+

Commit:
remove task executions, add extra phpunit test

- remove execution of tasks after 'fix' applied.
- added extra phpunit test for

bradpasley committed:
Merge pull request #16 from catalyst/dev-reschedule-not-execute

remove task executions, add extra phpunit test

Commit:
refactor class names & phpunit test class names

golenkovm reviewed (approved):
The code looks much better now. Thanks @bradpasley for making the changes.
